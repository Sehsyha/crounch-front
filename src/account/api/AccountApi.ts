import { Api } from '@/common/api/Api'
import { doFetch } from '@/api/doFetch'
import { LoginResponse, SignupResponse } from './repsonses'

export class AccountApi extends Api {
  protected BASE_URL = 'account'

  public signup(email: string, password: string): Promise<SignupResponse> {
    const url = this.computeUrl('signup')
    return doFetch({
      url: url,
      method: 'POST',
      data: {
        email,
        password,
      },
    });
  }

  public login(email: string, password: string): Promise<LoginResponse> {
    const url = this.computeUrl('login')

    return doFetch({
      url,
      method: 'POST',
      data: {
        email,
        password,
      },
    })
  }

  public logout(): Promise<void> {
    const url = this.computeUrl('logout')

    return doFetch({
      url,
      method: 'POST'
    })
  }
}
