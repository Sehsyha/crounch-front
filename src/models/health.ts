export interface Health {
  alive: boolean;
  version: string;
}
