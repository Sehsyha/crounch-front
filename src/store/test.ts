import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
