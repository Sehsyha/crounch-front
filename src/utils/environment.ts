export function getAPIURL(): string {
  return process.env.NODE_ENV === 'production' ? 'http://crounch.me:3000/' : 'http://localhost:3000/';
}
