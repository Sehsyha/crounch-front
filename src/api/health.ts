import { api } from './api';
import { Health } from '@/models/health';

export function health(): Promise<Health> {
  return api.get<Health>('health').then(res => res.data);
}
