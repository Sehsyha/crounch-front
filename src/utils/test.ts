import Vue from 'vue';
import VueRouter from 'vue-router';

import Vuex, { ActionContext, ActionTree, ModuleTree } from 'vuex';
import { shallowMount, VueClass, createLocalVue, Slots } from '@vue/test-utils';

import { RootState, ActionCall } from '@/store';

export interface ShallowOptions {
  values?: object;
  modules?: ModuleTree<RootState>;
  slots?: Slots;
}

const emptyOptions: ShallowOptions = {
  values: {},
  modules: {},
  slots: {}
};

export function shallowComponent<V extends Vue>(component: VueClass<V>, { values, modules, slots }: ShallowOptions = emptyOptions) {
  const localVue = createLocalVue();

  localVue.use(Vuex);

  const router = new VueRouter();

  const store = new Vuex.Store<RootState>({
    modules,
  });

  return shallowMount(component, {
    attachToDocument: true,
    propsData: {
      ...values,
    },
    slots,
    store,
    router,
    localVue,
    stubs: ['router-link', 'router-view'],
    mocks: {
      $router: {
        push: jest.fn()
      }
    }
  });
}

export const callAction = <S>(context: ActionContext<S, RootState>, actions: ActionTree<S, RootState>) => (
  key: string,
  params?: any
): any => {
  return (actions[key] as ActionCall<S, RootState>)(context, params);
};

export function initContext<S>(context: Partial<ActionContext<S, RootState>>): ActionContext<S, RootState> {
  const emptyContext = createEmptyActionContext<S>();

  return {
    ...emptyContext,
    ...context,
  };
}

function createEmptyActionContext<S>(): ActionContext<S, RootState> {
  return {
    state: {},
    getters: {},
    dispatch: {},
    commit: {},
    rootState: {
      dummy: '',
    },
    rootGetters: {},
  } as ActionContext<S, RootState>;
}
