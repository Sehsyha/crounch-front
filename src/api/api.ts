import { getAPIURL } from '@/utils/environment';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { FetchError } from '@/utils/error';
import { AuthModule } from '@/store/AuthModule';
import { getModule } from 'vuex-module-decorators';
import router from '@/router/router';

export interface FetchOptions {
  url: string;
  data?: object;
  method: string;
}

export function doFetch<T>(options: FetchOptions): Promise<T> {
  return fetch(getUrl(options.url), getFetchOptions(options))
    .then(res => {
      if (res.status >= 200 && res.status <= 299 && res.status !== 204) {
        return res.json()
      }
      if (res.status === 204) {
        return
      }
      if (res.status === 401) {
        const authModule = getModule(AuthModule)
        authModule.logoutAction().then(() => {
          router.push({ name: 'home' })
        })
      }

      return res.json()
        .then(body => { throw new FetchError('Erreur lors de la requÃªte', res.status, body) })
        .catch()
    })
}

function getFetchOptions({ data, method }: FetchOptions): object {
  let options: any = {
    method,
    headers: {
      Authorization: localStorage.getItem(TOKEN_STORAGE_KEY),
    },
  };

  if (method !== 'GET') {
    options.body = JSON.stringify(data);
  }

  return options;
}

function getUrl(url: string): string {
  return `${getAPIURL()}/${url}`;
}
