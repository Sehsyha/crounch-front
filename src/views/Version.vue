<template>
  <div>
    <h1>Version</h1>
    <p>
      Front:
      <span v-text="frontVersion" />
    </p>
    <p>
      Back:
      <span v-text="backVersion" />
    </p>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { health } from '@/api/health';
import { getVersion } from '@/utils/environment';

@Component
export default class Version extends Vue {
  backVersion: string = '';
  frontVersion: string = getVersion();

  mounted() {
    health().then(res => (this.backVersion = res.version));
  }
}
</script>
