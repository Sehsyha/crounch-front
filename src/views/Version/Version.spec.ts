import Version from './Version.vue';
import { health } from '@/api/health';
import { getVersion } from '@/utils/environment';
import { shallowComponent } from '@/utils/test';

jest.mock('@/api/health');
jest.mock('@/utils/environment');

describe('Version', () => {
  beforeEach(() => {
    (health as jest.Mock).mockResolvedValue({ version: '1.2.3' });
    (getVersion as jest.Mock).mockReturnValue('3.2.1');
  });

  describe('Init', () => {
    it('Should initialize backVersion to empty string.', () => {
      const wrapper = shallowComponent(Version);

      expect(wrapper.vm.$data.backVersion).toBe('');
    });

    it('Should initialize frontVersion to current version.', () => {
      const wrapper = shallowComponent(Version);

      expect(wrapper.vm.$data.frontVersion).toBe('3.2.1');
    });
  });

  it('Should render.', done => {
    const wrapper = shallowComponent(Version);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
