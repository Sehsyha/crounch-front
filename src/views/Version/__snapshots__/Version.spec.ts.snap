// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Version Should render. 1`] = `
<div>
  <h1>Version</h1>
  <p>
    Front:
    <span>3.2.1</span></p>
  <p>
    Back:
    <span>1.2.3</span></p>
</div>
`;
